### PR TITLE
adding completion time for job failed due to exceeding deadline

### LINKED ISSUE #86436: Kubectl reporting wrong duration for jobs aborted due to exceeding deadline
**What happened**:

It seems that Jobs that get terminated due to exceeding the active deadline seconds get a wrong duration reported via `kubectl`.

This the `kubectl get jobs` for the affected Jobs:

```
$ kubectl get jobs -l app=material-girl                              
NAME                                          COMPLETIONS   DURATION   AGE
material-girl-dashboard-tracking-1575862380   1/1           4m2s       10d
material-girl-dashboard-tracking-1575948780   1/1           3m18s      9d
material-girl-dashboard-tracking-1576035180   0/1           8d         8d
material-girl-dashboard-tracking-1576121580   1/1           5m16s      7d10h
material-girl-dashboard-tracking-1576207980   1/1           4m12s      6d10h
material-girl-dashboard-tracking-1576294380   1/1           3m26s      5d10h
material-girl-dashboard-tracking-1576380780   0/1           4d10h      4d10h
material-girl-dashboard-tracking-1576467180   0/1           3d10h      3d10h <-- This
material-girl-dashboard-tracking-1576481580   0/1           3d1h       3d1h
material-girl-dashboard-tracking-1576567980   0/1           2d6h       2d6h
material-girl-dashboard-tracking-1576654380   0/1           30h        30h
material-girl-dashboard-tracking-1576740780   1/1           3m43s      6h28m
```

There are other Jobs affected by this issue (all of those created via a Cronjob), but let me show one particular example.

Running `kubectl describe` don't give any information about the status of this Job, as it would if it was still running, I think, which is also not expected behavior:

```
Name:                     material-girl-dashboard-tracking-1576467180
Namespace:                default
Selector:                 controller-uid=c7573657-1fb4-11ea-bb06-42010af0026d
Labels:                   app=material-girl
                          bu=information
                          name=material-girl-dashboard-tracking
Annotations:              <none>
Controlled By:            CronJob/material-girl-dashboard-tracking
Parallelism:              1
Completions:              1
Start Time:               Mon, 16 Dec 2019 00:33:08 -0300
Active Deadline Seconds:  900s
Pods Statuses:            0 Running / 0 Succeeded / 1 Failed
Pod Template:
  Labels:  app=material-girl
           bu=information
           controller-uid=c7573657-1fb4-11ea-bb06-42010af0026d
           job-name=material-girl-dashboard-tracking-1576467180
           name=material-girl-dashboard-tracking
  Containers:
   material-girl-dashboard-tracking:
    Image:      redacted
    Port:       <none>
    Host Port:  <none>
    Command:
      redacted
    Limits:
      memory:  128M
    Requests:
      cpu:     200m
      memory:  128M
    Environment Variables from:
      redacted
    Environment:
      redacted
    Mounts:           <none>
  Volumes:            <none>
Events:               <none>
```

This is the YAML manifest for that Job:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  creationTimestamp: "2019-12-16T03:33:08Z"
  labels:
    app: material-girl
    bu: information
    name: material-girl-dashboard-tracking
  name: material-girl-dashboard-tracking-1576467180
  namespace: default
  ownerReferences:
  - apiVersion: batch/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: CronJob
    name: material-girl-dashboard-tracking
    uid: 00f35158-1041-11ea-948c-42010af002a4
  resourceVersion: "1343281381"
  selfLink: /apis/batch/v1/namespaces/default/jobs/material-girl-dashboard-tracking-1576467180
  uid: c7573657-1fb4-11ea-bb06-42010af0026d
spec:
  activeDeadlineSeconds: 900
  backoffLimit: 1
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      controller-uid: c7573657-1fb4-11ea-bb06-42010af0026d
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: material-girl
        bu: information
        controller-uid: c7573657-1fb4-11ea-bb06-42010af0026d
        job-name: material-girl-dashboard-tracking-1576467180
        name: material-girl-dashboard-tracking
      name: material-girl-dashboard-tracking
    spec:
      containers:
      - command:
        - redacted
        env:
        - name: redacted
          value: redacted
        image: redacted
        imagePullPolicy: Always
        name: material-girl-dashboard-tracking
        resources:
          limits:
            memory: 128M
          requests:
            cpu: 200m
            memory: 128M
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Never
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  conditions:
  - lastProbeTime: "2019-12-16T03:48:08Z"
    lastTransitionTime: "2019-12-16T03:48:08Z"
    message: Job was active longer than specified deadline
    reason: DeadlineExceeded
    status: "True"
    type: Failed
  failed: 1
  startTime: "2019-12-16T03:33:08Z"
```

Kubectl is reporting the correct duration for Jobs that finished successfully. The only difference is that successful Jobs have a `completionTime` status field that is not being set when the Job aborts due to reaching the active deadline seconds, so this might not be a problem in `kubectl` itself.

This is the status of a Job that succeeded:

```yaml
status:
  completionTime: "2019-12-12T03:38:24Z"
  conditions:
  - lastProbeTime: "2019-12-12T03:38:24Z"
    lastTransitionTime: "2019-12-12T03:38:24Z"
    status: "True"
    type: Complete
  startTime: "2019-12-12T03:33:08Z"
  succeeded: 1
```

**What you expected to happen**:

I expected `kubectl get jobs` to report the correct job duration when job was aborted due to exceeding the active deadline seconds.

I expected `kubectl describe job` to display the status for Jobs aborted due to exceeding the active deadline seconds.

I expected Kubernetes to set the `completionTime` status field for aborted Jobs, maybe?

**How to reproduce it (as minimally and precisely as possible)**:

See YAML files above, they should be enough to reproduce the issue.

**Anything else we need to know?**:

**Environment**: GKE

- Kubernetes version (use `kubectl version`):
  - Server: 1.14.8-gke.12
  - Client: 1.16.3

### PR DESCRIPTION
**Pull request by eladm26**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
it addresses the issue where job's duration doesn't show the correct time when job exceeded the deadline

**Which issue(s) this PR fixes**:
Fixes #86436 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

### CODE REVIEW

**Comment by eladm26** [comment]
sig/cli

**Review by moret — APPROVED** [review]
LGTM

**Comment by brianpursley** [comment]
I also ran into the issue associated with this PR today.

@eladm26 it seems like a pretty small change, but is there a unit test associated with this code where you can add a case?

/cc @janetkuo

**Review by janetkuo — CHANGES_REQUESTED** [review]
A Job can be either `Complete` or `Failed`, but not both. 
`CompletionTime` is only set when the Job is complete, i.e. when a specified number of successful completions is reached, 
https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/

**Comment by janetkuo** [comment]
@eladm26 thanks for your contribution. Here are some changes we need instead. 

* Here's an issue about how we can improve API docs: #68712
* Make the doc more clear, too https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/
* Add a field that records the time when the Job fails

**Comment by eladm26** [comment]
No problem, will have a look.

**Comment by eladm26** [comment]
@janetkuo updated the PR

File: pkg/apis/batch/types.go
Comment (soltysh) [inline pkg/apis/batch/types.go]: Please do not extend the current API, it's sufficient as is. You should just ensure to fill in `CompletionTime` and the job status will properly reflect that it failed due to deadline.

File: pkg/controller/job/job_controller.go
Comment (soltysh) [inline pkg/controller/job/job_controller.go]: Don't, set `CompletionTime`, instead.

File: pkg/printers/internalversion/printers.go
Comment (soltysh) [inline pkg/printers/internalversion/printers.go]: Not needed.

**Review by soltysh — CHANGES_REQUESTED** [review]
/hold

File: pkg/apis/batch/types.go
Comment (eladm26) [reply pkg/apis/batch/types.go]: @soltysh thank you for the review, that was actually my first fix, but I added the `FailureTime` field per @janetkuo request:
> @eladm26 thanks for your contribution. Here are some changes we need instead.
> 
>     * Here's an issue about how we can improve API docs: #68712
> 
>     * Make the user guilde more clear, too https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/
> 
>     * Add an API field in Job that records the time when the Job fails

so just to be clear, is that the final decision?

**Comment by janetkuo** [comment]
@soltysh completion time is only set when the Job completes successfully. A Failed Job isn't Complete. Setting it when a Job fails breaks compatibility.

**Comment by eladm26** [comment]
@soltysh @janetkuo Hi, sorry to bother but is there any update regarding the appropriate solution?

**Comment by loganmzz** [comment]
I suggest to use `terminationTime` (or similar) that could match both cases. Is there any state schema/diagram that can ensure right term usage ?

**Comment by t0rr3sp3dr0** [comment]
It would be nice to have this merged by Kubernetes v1.19

**Comment by eladm26** [comment]
@janetkuo what did we decide about this one?

**Comment by janetkuo** [comment]
@soltysh any further thoughts on this one? Job has 2 different conditions: `Complete` and `Failed`. It'll be confusing (and breaking) to set `CompletionTime` when a Job fails.

**Comment by liggitt** [comment]
[Before this gets into API review](https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#mechanics), make sure the goal of the proposed change has agreement from the appropriate sub-project owners and/or SIG leads. It sounds like more discussion was needed with @soltysh and @janetkuo

**Review by soltysh — APPROVED** [review]
@eladm26 I'm finally returning to this one, and I was wrong. This PR looks good from my perspective. Please get @liggitt approval on the API and this is good to go.
/lgtm
/approve
/ok-to-test
/hold cancel

**Review by liggitt — COMMENTED** [review]
The need for a new field to address #86436 is not clear to me. It seems like the information is already present in the lastTransitionTime field of the Failed condition. If the primary goal is to make the printed output of `kubectl get` nicer in this scenario, I would start by using that information over adding a new API field.

**Comment by liggitt** [comment]
see https://github.com/liggitt/kubernetes/commits/job-failed-time as an alternative, using the data already persisted in the API

**Pull request closed without merging**
